### DISCUSSION TITLE
Raising custom exceptions subclassing `ValueError` from validator

### DISCUSSION THREAD

**Discussion by lincior** [pydantic/pydantic#8563] [Question] [answered]
Is it possible to define a `pydantic.BaseModel`, with a custom validator raising an exception subclassing `ValueError`?
```python
from pydantic import model_validator, BaseModel, ValidationError

class MyValueError(ValueError):
    pass

class CustomModel(BaseModel):
    dummy: str

    @model_validator(mode="after")
    def dummy_validator(self, v):
        print("should raise MyException")
        raise MyValueError("dummy_validator")

try:
    CustomModel(dummy="")
except MyValueError:
    print("MyValueError caught")
except ValidationError:
    print("pydantic.ValidationError caught")
```
Output:
```
should raise MyException
pydantic.ValidationError caught
```
I noticed that it works if instead of extending `ValueError`, the exception class extends `Exception`, but this is not my ideal case:
```python
from pydantic import model_validator, BaseModel, ValidationError

class MyException(Exception):
    pass

class CustomModel(BaseModel):
    dummy: str

    @model_validator(mode="after")
    def dummy_validator(self, v):
        if v != "dummy":
            print("should raise MyException")
            raise MyException("dummy_validator")

try:
    CustomModel(dummy="")
except MyException:
    print("MyException caught")
except ValidationError:
    print("pydantic.ValidationError caught")
```
Outputs:
```
should raise MyException
MyException caught
```

**Comment by hramezani** [comment, marked as answer]
I think it's only `ValidationError` that will be raised by pydantic and you can catch it in the `try/except` block.

I think you can use `PydanticCustomError`. take a look at the [example in our doc](https://docs.pydantic.dev/latest/concepts/validators/#handling-errors-in-validators).

Also, checking [custom error doc](https://docs.pydantic.dev/latest/errors/errors/#custom-errors) probably helps you.

**Reply by lincior** [reply]
Thanks, this explains the observed behaviour.

**Reply by OluwaninsolaAO** [reply]
@RiccardoLincetto @hramezani 
Following the conversation here, I may be facing a somewhat related issue with a project. Basically, I cannot catch `pydantic.ValidationError` or `ValueError` separately in a single `try - except` block, the first one will automatically be a match in the `except` block.

```python
    try:
        # ... raises pydantic.ValidationError due to failed validation
        pass
    except ValueError as exc:
        abort(400, description=str(exc))  # ValueError will be matched
    except pydantic.ValidationError as exc:
        abort(403, description=str(exc))
```
or
```python
    try:
        # ... raises ValueError due to some forbidden action
        pass
    except pydantic.ValidationError as exc:
        abort(403, description=str(exc))  # pydantic.ValidationError will be matched
    except ValueError as exc:
        abort(400, description=str(exc))
```

This is concerning, I couldn't handle pydantic validation error separately from naitive python `ValueError`. Any suggestions?
